### ISSUE TITLE
BUG: modin on ray produce error with empty dataframes

### ISSUE THREAD

**Issue by Egor-Krivov** [modin-project/modin#5430]
### Modin version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the latest released version of Modin.

- [X] I have confirmed this bug exists on the main branch of Modin. (In order to do this you can follow [this guide](https://modin.readthedocs.io/en/stable/getting_started/installation.html#installing-from-the-github-master-branch).)

### Reproducible Example

```python
pd.DataFrame({'item': [1, 2], 'week': [1, 2]}).query('10 < week').groupby('item').size()
```

### Issue Description

This code produces an error with modin on ray. However, works on pandas. I have a bit different code in my benchmark and it just works on pandas but fails on modin.

### Expected Behavior

Modin on ray should behave like pandas.

### Error Logs

<details>

```
(MODIN-ON-RAY)
pd.DataFrame({'item': [1, 2], 'week': [1, 2]}).query('10 < week').groupby('item').size()
UserWarning: Distributing <class 'dict'> object. This may take some time. 
UserWarning: `DataFrame.__getitem__` for empty DataFrame is not currently supported by PandasOnRay, defaulting to pandas implementation.
UserWarning: Distributing <class 'pandas.core.series.Series'> object. This may take some time.
*** IndexError: list index out of range 

(PANDAS) 
>>> pd.DataFrame({'item': [1, 2], 'week': [1, 2]}).query('10 < week').groupby('item').size()
Series([], dtype: int64)
```
</details>

### Installed Versions

<details>

UserWarning: Setuptools is replacing distutils.

INSTALLED VERSIONS
------------------
commit           : c30ab4c132295b1e168fbb06a050c76554759e11
python           : 3.8.15.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.4.0-124-generic
Version          : #140-Ubuntu SMP Thu Aug 4 02:23:37 UTC 2022
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US
LOCALE           : en_US.ISO8859-1

Modin dependencies
------------------
modin            : 0.7.3+1359.gc30ab4c1
ray              : 2.0.1
dask             : None
distributed      : None
hdk              : present

pandas dependencies
-------------------
pandas           : 1.5.1
numpy            : 1.23.4
pytz             : 2022.6
dateutil         : 2.8.2
setuptools       : 65.5.0
pip              : 22.2.2
Cython           : 0.29.32
pytest           : 7.2.0
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : 8.6.0
pandas_datareader: None
bs4              : None
bottleneck       : 1.3.5
brotli           :
fastparquet      : None
fsspec           : 2022.11.0
gcsfs            : None
matplotlib       : 3.6.2
numba            : None
numexpr          : 2.8.4
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : 6.0.0
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : 1.9.3
snappy           : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
zstandard        : None
tzdata           : None

</details>

**Comment by Egor-Krivov** [comment]
Surprisingly, in my case this problem appears with non-empty array after `query`. And for some reason doing `.iloc[:10000000000000]` right after query solves the issue.

**Comment by vnlitvinov** [comment]
@Egor-Krivov you seem to have quite a funny Modin version reported, how did you install that?

cc @dchigarev I wonder if your recent `groupby().size()` PR fixed that...

I've run the reproducer on my Windows laptop, and I'm seeing a different error here:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Vass\ponder\modin\modin\logging\logger_decorator.py", line 128, in run_and_log
    return obj(*args, **kwargs)
  File "C:\Vass\ponder\modin\modin\pandas\dataframe.py", line 480, in groupby
    if by is not None and by in self._query_compiler.get_index_names(axis):
  File "C:\Vass\ponder\modin\modin\logging\logger_decorator.py", line 128, in run_and_log
    return obj(*args, **kwargs)
  File "C:\Vass\ponder\modin\modin\core\storage_formats\base\query_compiler.py", line 3464, in get_index_names
    return self.get_axis(axis).names
AttributeError: 'list' object has no attribute 'names'
```

**Comment by Egor-Krivov** [comment]
Indeed, my version is strange. I will reproduce my installation and check what will change after that.

**Comment by dchigarev** [comment]
There are two distinct problems causing this behavior that I found so far (hope I won't find more :D):
1. When Modin's logic produces an empty dataframe (e.g. filtering) it writes an empty list into the frame's `.index` attribute instead of an empty `pandas.Index` object. This obviously causes problems when trying to treat a `df.index` as an actual index:
    ```python
    >>> import modin.pandas as pd
    >>> res = pd.DataFrame({"a": [1, 2, 3]}).query("a > 200")
    >>> res.index
    []
    >>> type(res.index)
    <class 'list'>
    >>> res.index.name
    Traceback (most recent call last):
      File "<stdin>", line 1, in <module>
    AttributeError: 'list' object has no attribute 'name'
    ```
    This happens because the logic that computes indices unconditionally filters out empty ones, thus returning an empty list instead of an index in case all of the partitions returned a zero-length index.
    https://github.com/modin-project/modin/blob/4114183f84d3d152126510655db5698fd4ea1071/modin/core/dataframe/pandas/partitioning/partition_manager.py#L883-L888
    I've fixed this exact problem in #5431
2. Modin falls down to an actual groupby execution on an empty frame in case an underlying query compiler is lazy (e.g. has delayed indices). Groupby implementation was not developed with the intention to work on empty frames (it was believed that empty frame cases would always default to pandas), however, in the reproducer, we see the case when an empty dataframe snuck into low-level implementation and breaks everything. Here's the scenario of how this happens:
   a. `.query` [sets resulting index-cache to `None`](https://github.com/modin-project/modin/blob/4114183f84d3d152126510655db5698fd4ea1071/modin/core/dataframe/pandas/dataframe/dataframe.py#L2113) which causes the query compiler [to be lazy](https://github.com/modin-project/modin/blob/4114183f84d3d152126510655db5698fd4ea1071/modin/core/storage_formats/pandas/query_compiler.py#L269-L282).
    b. When next time checking for a posibility to default to pandas on an empty frame Modin skips this check as the frame has delayed executions [[1]](https://github.com/modin-project/modin/blob/4114183f84d3d152126510655db5698fd4ea1071/modin/pandas/base.py#L3918)
   c. We end-up in a groupby execiton on an empty frame.
